# <span style='font-size:50px;background-color:yellow;'>&#128201;</span> <font color="#004080"> *Sentiment analysis on IMDB movie review dataset ::* </font>

## The dataset comprises of 50,000 movie reviews taken from IMDb. The training and test data sets are uploaded in moodle. The datasets are in the form of text files. Every line in the files is a movie review. The training set and test set comprises of 25000 movie reviews each. For both training and test data sets, the first 12500 reviews are positive and the rest of the reviews are negative.

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

In [3]:
from google.colab import files
files.upload()

Saving test.txt to test.txt
Saving train.txt to train.txt


In [0]:
train_data = open("train.txt").read()
train_reviews = train_data.split("\n")

test_data = open("test.txt").read()
test_reviews = test_data.split("\n")

print(train_reviews[0])

Bromwell High is a cartoon comedy. It ran at the same time as some other programs about school life, such as "Teachers". My 35 years in the teaching profession lead me to believe that Bromwell High's satire is much closer to reality than is "Teachers". The scramble to survive financially, the insightful students who can see right through their pathetic teachers' pomp, the pettiness of the whole situation, all remind me of the schools I knew and their students. When I saw the episode in which a student repeatedly tried to burn down the school, I immediately recalled ......... at .......... High. A classic line: INSPECTOR: I'm here to sack one of your teachers. STUDENT: Welcome to Bromwell High. I expect that many adults of my age think that Bromwell High is far fetched. What a pity that it isn't!


### (a) Preprocess the dataset: Remove punctuation and < br > tag. Convert all the characters to lower case --> 

In [0]:
from string import punctuation
print(punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [0]:
# for train data

# removing forms of <br>
for i, element in enumerate(train_reviews):
    element = element.replace('<br>',' ')
    element = element.replace('</br>',' ')
    element = element.replace('<br />',' ')
    train_reviews[i] = element

# removing punctuations & lower the case
for i, element in enumerate(train_reviews):
  s = ''
  for charac in element:
    if charac not in punctuation:
      s += charac.lower()
  train_reviews[i] = s 

print(train_reviews[0])

bromwell high is a cartoon comedy it ran at the same time as some other programs about school life such as teachers my 35 years in the teaching profession lead me to believe that bromwell highs satire is much closer to reality than is teachers the scramble to survive financially the insightful students who can see right through their pathetic teachers pomp the pettiness of the whole situation all remind me of the schools i knew and their students when i saw the episode in which a student repeatedly tried to burn down the school i immediately recalled  at  high a classic line inspector im here to sack one of your teachers student welcome to bromwell high i expect that many adults of my age think that bromwell high is far fetched what a pity that it isnt


In [0]:
# for test data

# removing forms of <br>
for i, element in enumerate(test_reviews):
    element = element.replace('<br>',' ')
    element = element.replace('</br>',' ')
    element = element.replace('<br />',' ')
    test_reviews[i] = element

# removing punctuations & lower the case
for i, element in enumerate(test_reviews):
  s = ''
  for charac in element:
    if charac not in punctuation:
      s += charac.lower()
  test_reviews[i] = s 

print(test_reviews[0])

i went and saw this movie last night after being coaxed to by a few friends of mine ill admit that i was reluctant to see it because from what i knew of ashton kutcher he was only able to do comedy i was wrong kutcher played the character of jake fischer very well and kevin costner played ben randall with such professionalism the sign of a good movie is that it can toy with our emotions this one did exactly that the entire theater which was sold out was overcome by laughter during the first half of the movie and were moved to tears during the second half while exiting the theater i not only saw many women in tears but many full grown men as well trying desperately not to let anyone see them crying this movie was great and i suggest that you go see it before you judge


### (b) Create the target vector for training and test data. The first 12500 values of target vector is 1 and the rest of the values are zero. The target vector is same for training and test data -->

In [0]:
# removing last empty review created after splitting with '\n'
train_reviews = train_reviews[:25000]
test_reviews = test_reviews[:25000]
print(len(train_reviews))

25000


In [0]:
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences

Using TensorFlow backend.


In [0]:
tokenizer = Tokenizer(num_words = 5000)
tokenizer.fit_on_texts(train_reviews)

train_reviews_tokenized = tokenizer.texts_to_sequences(train_reviews)      

print(train_reviews[0])
print(train_reviews_tokenized[0])

bromwell high is a cartoon comedy it ran at the same time as some other programs about school life such as teachers my 35 years in the teaching profession lead me to believe that bromwell highs satire is much closer to reality than is teachers the scramble to survive financially the insightful students who can see right through their pathetic teachers pomp the pettiness of the whole situation all remind me of the schools i knew and their students when i saw the episode in which a student repeatedly tried to burn down the school i immediately recalled  at  high a classic line inspector im here to sack one of your teachers student welcome to bromwell high i expect that many adults of my age think that bromwell high is far fetched what a pity that it isnt
[321, 6, 3, 1073, 216, 8, 2083, 29, 1, 165, 58, 13, 45, 78, 41, 398, 116, 135, 13, 4884, 54, 4934, 146, 7, 1, 4935, 479, 68, 5, 254, 11, 1958, 6, 71, 2373, 5, 629, 69, 6, 4884, 1, 5, 1964, 1, 1431, 35, 67, 65, 202, 139, 63, 1191, 4884, 1

In [0]:
#test_tokenizer = Tokenizer(num_words = 5000)
#tokenizer.fit_on_texts(test_reviews)

test_reviews_tokenized = tokenizer.texts_to_sequences(test_reviews)      

print(test_reviews[0])
print(test_reviews_tokenized[0])
print(type(test_reviews_tokenized))

i went and saw this movie last night after being coaxed to by a few friends of mine ill admit that i was reluctant to see it because from what i knew of ashton kutcher he was only able to do comedy i was wrong kutcher played the character of jake fischer very well and kevin costner played ben randall with such professionalism the sign of a good movie is that it can toy with our emotions this one did exactly that the entire theater which was sold out was overcome by laughter during the first half of the movie and were moved to tears during the second half while exiting the theater i not only saw many women in tears but many full grown men as well trying desperately not to let anyone see them crying this movie was great and i suggest that you go see it before you judge
[9, 421, 2, 208, 10, 16, 226, 308, 99, 107, 5, 31, 3, 162, 338, 4, 1887, 529, 942, 11, 9, 12, 5, 65, 8, 82, 34, 47, 9, 678, 4, 26, 12, 60, 494, 5, 81, 216, 9, 12, 351, 250, 1, 106, 4, 3237, 51, 72, 2, 1797, 250, 1002, 15, 

In [0]:
# target vector for training, and test data

a=[1]*12500
b=[0]*12500
train_target_vec = a + b
test_target_vec = a + b

### (c) Create a validation set which is 20% of the training set -->

In [0]:
#shuffling the training dataset and creating validation dataset
#we tend to maintain the same ratio of class = 1 , class = 0, members in train and validation data

# train_data_size = 25000
# train_data_size_of_each_class = 25000
# 20/100 * 12500 = 2500

# shuffling class=1 data indices
new_indices1 = np.arange(0, 12500)
np.random.shuffle(new_indices1)

#shuffling class=0 data indices
new_indices0 = np.arange(12500, 25000)
np.random.shuffle(new_indices0)

#dividing training into classes
#train_reviews1 = train_reviews[0:12500]
train_reviews_tokenized1 = train_reviews_tokenized[0:12500]
#train_reviews0 = train_reviews[12500:25000]
train_reviews_tokenized0 = train_reviews_tokenized[12500:25000]

#print(type(new_indices1))

#shuffling data, each class separately
#X_new1 = train_reviews1[new_indices1]
train_reviews_tokenized1 = np.array(train_reviews_tokenized1)
train_reviews_tokenized0 = np.array(train_reviews_tokenized0)
X_new_tok1 = train_reviews_tokenized1[new_indices1]
#X_new0 = train_reviews0[new_indices0 - 12500]
X_new_tok0 = train_reviews_tokenized0[new_indices0 - 12500]

# creating x_val and x_train
X_val = np.array( list(X_new_tok1[0:2500]) + list(X_new_tok0[0:2500]) )
X_train = np.array( list(X_new_tok1[2500:]) + list(X_new_tok0[2500:]) )
Y_val = np.array( list([1]*2500) + list([0]*2500) )
Y_train = np.array( list([1]*10000) + list([0]*10000) )

X_test = np.array(test_reviews_tokenized[:])
Y_test = np.array(test_target_vec[:])

#Shuffling train, test, validation datasets independently
new_indices = np.arange(0, X_train.shape[0])
np.random.shuffle(new_indices)
X_train = X_train[new_indices]
Y_train = Y_train[new_indices]

new_indices = np.arange(0, X_test.shape[0])
np.random.shuffle(new_indices)
X_test = X_test[new_indices]
Y_test = Y_test[new_indices]

new_indices = np.arange(0, X_val.shape[0])
np.random.shuffle(new_indices)
X_val = X_val[new_indices]
Y_val = Y_val[new_indices]

#printing sizes
print("Final shapes -->")
print("Shape of X_train : {}".format(np.shape(X_train)))
print("Shape of Y_train : {}".format(np.shape(Y_train)))
print("Shape of X_test : {}".format(np.shape(X_test)))
print("Shape of Y_test : {}".format(np.shape(Y_test)))
print("Shape of X_val : {}".format(np.shape(X_val)))
print("Shape of Y_val : {}".format(np.shape(Y_val)))

Final shapes -->
Shape of X_train : (20000,)
Shape of Y_train : (20000,)
Shape of X_test : (25000,)
Shape of Y_test : (25000,)
Shape of X_val : (5000,)
Shape of Y_val : (5000,)


In [0]:
print(Y_train)

[0 0 1 ... 0 1 0]


### (d) Train the following models -->

Integer encode the reviews, is done in section (c) 

In [0]:
from tensorflow.keras import layers
from tensorflow.keras.layers import Dense, Dropout, Activation, Flatten
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Embedding, Conv1D, GRU, LSTM, MaxPool1D
from tensorflow.keras.layers import RNN, SimpleRNN, Bidirectional

### (e) Model 1 -->

In [0]:
X_train1 = pad_sequences(X_train, maxlen = 200)   
X_test1 = pad_sequences(X_test, maxlen = 200)   
X_val1 = pad_sequences(X_val, maxlen = 200)   

In [0]:
model1 = tf.keras.Sequential([
    Embedding(5000 ,128 , input_length = 200),
    SimpleRNN(200, activation="tanh"),
    Dense(1, activation='sigmoid')
])
print (model1.summary())

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_10 (Embedding)     (None, 200, 128)          640000    
_________________________________________________________________
simple_rnn_5 (SimpleRNN)     (None, 200)               65800     
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 201       
Total params: 706,001
Trainable params: 706,001
Non-trainable params: 0
_________________________________________________________________
None


In [0]:
iter=15
opt = tf.keras.optimizers.Adam(0.0002)
model1.compile(loss='binary_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])

epoc1 = model1.fit(X_train1.astype(np.float32), Y_train, epochs = iter, verbose=2)

Train on 20000 samples
Epoch 1/15
20000/20000 - 58s - loss: 0.6808 - acc: 0.5526
Epoch 2/15
20000/20000 - 56s - loss: 0.4962 - acc: 0.7656
Epoch 3/15
20000/20000 - 56s - loss: 0.3581 - acc: 0.8509
Epoch 4/15
20000/20000 - 56s - loss: 0.1773 - acc: 0.9345
Epoch 5/15
20000/20000 - 56s - loss: 0.0802 - acc: 0.9744
Epoch 6/15
20000/20000 - 56s - loss: 0.0353 - acc: 0.9921
Epoch 7/15
20000/20000 - 56s - loss: 0.0089 - acc: 0.9989
Epoch 8/15
20000/20000 - 56s - loss: 0.0021 - acc: 1.0000
Epoch 9/15
20000/20000 - 56s - loss: 8.7989e-04 - acc: 1.0000
Epoch 10/15
20000/20000 - 56s - loss: 4.7818e-04 - acc: 1.0000
Epoch 11/15
20000/20000 - 56s - loss: 3.0067e-04 - acc: 1.0000
Epoch 12/15
20000/20000 - 57s - loss: 1.9823e-04 - acc: 1.0000
Epoch 13/15
20000/20000 - 55s - loss: 1.3225e-04 - acc: 1.0000
Epoch 14/15
20000/20000 - 56s - loss: 9.0450e-05 - acc: 1.0000
Epoch 15/15
20000/20000 - 57s - loss: 6.2060e-05 - acc: 1.0000


In [0]:
# RNN
lo, ac = model1.evaluate(X_train1.astype(np.float32), Y_train)
print("Train Loss: {}".format(lo))
print("Train Accuracy: {}".format(ac)) 

lo, ac = model1.evaluate(X_val1.astype(np.float32), Y_val)
print("Validation Loss: {}".format(lo))
print("Validation Accuracy: {}".format(ac)) 

lo, ac = model1.evaluate(X_test1.astype(np.float32), Y_test)
print("Test Loss: {}".format(lo))
print("Test Accuracy: {}".format(ac)) 

20000/20000 [==============================] - 12s 582us/sample - loss: 4.8227e-05 - acc: 1.0000
Train Loss: 4.8226801701821386e-05
Train Accuracy: 1.0
5000/5000 [==============================] - 3s 570us/sample - loss: 0.8579 - acc: 0.8320
Validation Loss: 0.8579156356334686
Validation Accuracy: 0.8320000171661377
25000/25000 [==============================] - 14s 572us/sample - loss: 0.8692 - acc: 0.8352
Test Loss: 0.8691968902683258
Test Accuracy: 0.8352000117301941


### (f) Model 2 -->

In [0]:
model2 = tf.keras.Sequential([
    Embedding(5000 ,128 , input_length = 200),
    LSTM(200, activation="tanh"),
    Dense(1, activation='sigmoid')
])
print (model2.summary())

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_11 (Embedding)     (None, 200, 128)          640000    
_________________________________________________________________
lstm (LSTM)                  (None, 200)               263200    
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 201       
Total params: 903,401
Trainable params: 903,401
Non-trainable params: 0
_________________________________________________________________
None


In [0]:
iter=15
opt = tf.keras.optimizers.Adam(0.0002)
model2.compile(loss='binary_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])

epoc2 = model2.fit(X_train1.astype(np.float32), Y_train, epochs = iter, verbose=2)

Train on 20000 samples
Epoch 1/15
20000/20000 - 224s - loss: 0.4630 - acc: 0.7831
Epoch 2/15
20000/20000 - 226s - loss: 0.2695 - acc: 0.8933
Epoch 3/15
20000/20000 - 227s - loss: 0.2207 - acc: 0.9151
Epoch 4/15
20000/20000 - 228s - loss: 0.1959 - acc: 0.9280
Epoch 5/15
20000/20000 - 226s - loss: 0.1732 - acc: 0.9348
Epoch 6/15
20000/20000 - 225s - loss: 0.1590 - acc: 0.9438
Epoch 7/15
20000/20000 - 223s - loss: 0.1361 - acc: 0.9513
Epoch 8/15
20000/20000 - 221s - loss: 0.1238 - acc: 0.9567
Epoch 9/15
20000/20000 - 222s - loss: 0.1097 - acc: 0.9617
Epoch 10/15
20000/20000 - 224s - loss: 0.0915 - acc: 0.9712
Epoch 11/15
20000/20000 - 222s - loss: 0.0768 - acc: 0.9759
Epoch 12/15
20000/20000 - 222s - loss: 0.0666 - acc: 0.9801
Epoch 13/15
20000/20000 - 220s - loss: 0.0733 - acc: 0.9749
Epoch 14/15
20000/20000 - 222s - loss: 0.0486 - acc: 0.9850
Epoch 15/15
20000/20000 - 220s - loss: 0.0497 - acc: 0.9855


In [0]:
# LSTM
lo, ac = model2.evaluate(X_train1.astype(np.float32), Y_train)
print("Train Loss: {}".format(lo))
print("Train Accuracy: {}".format(ac)) 

lo, ac = model2.evaluate(X_val1.astype(np.float32), Y_val)
print("Validation Loss: {}".format(lo))
print("Validation Accuracy: {}".format(ac)) 

lo, ac = model2.evaluate(X_test1.astype(np.float32), Y_test)
print("Test Loss: {}".format(lo))
print("Test Accuracy: {}".format(ac)) 

20000/20000 [==============================] - 58s 3ms/sample - loss: 0.0310 - acc: 0.9937
Train Loss: 0.03098615928851068
Train Accuracy: 0.9936500191688538
5000/5000 [==============================] - 14s 3ms/sample - loss: 0.5896 - acc: 0.8596
Validation Loss: 0.5895959486365319
Validation Accuracy: 0.8596000075340271
25000/25000 [==============================] - 72s 3ms/sample - loss: 0.6065 - acc: 0.8484
Test Loss: 0.6065459573078156
Test Accuracy: 0.848360002040863


### (g) Model 3 -->

In [0]:
model3 = tf.keras.Sequential([
    Embedding(5000 ,128 , input_length = 200),
    GRU(200, activation="relu"),
    Dense(1, activation='sigmoid')
])
print (model3.summary())

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_13 (Embedding)     (None, 200, 128)          640000    
_________________________________________________________________
gru_1 (GRU)                  (None, 200)               197400    
_________________________________________________________________
dense_10 (Dense)             (None, 1)                 201       
Total params: 837,601
Trainable params: 837,601
Non-trainable params: 0
_________________________________________________________________
None


In [0]:
iter=15
opt = tf.keras.optimizers.Adam(0.0002)
model3.compile(loss='binary_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])

epoc3 = model3.fit(X_train1.astype(np.float32), Y_train, epochs = iter, verbose=2)

Train on 20000 samples
Epoch 1/15
20000/20000 - 194s - loss: 641581822.5923 - acc: 0.6148
Epoch 2/15
20000/20000 - 192s - loss: 0.5407 - acc: 0.7416
Epoch 3/15
20000/20000 - 194s - loss: 35562.9588 - acc: 0.7821
Epoch 4/15
20000/20000 - 193s - loss: 0.5065 - acc: 0.7978
Epoch 5/15
20000/20000 - 194s - loss: 0.4832 - acc: 0.8067
Epoch 6/15
20000/20000 - 194s - loss: 0.4627 - acc: 0.8127
Epoch 7/15
20000/20000 - 194s - loss: 0.4444 - acc: 0.8184
Epoch 8/15
20000/20000 - 195s - loss: 0.4281 - acc: 0.8235
Epoch 9/15
20000/20000 - 196s - loss: 0.4137 - acc: 0.8272
Epoch 10/15
20000/20000 - 197s - loss: 0.4010 - acc: 0.8333
Epoch 11/15
20000/20000 - 196s - loss: 0.3898 - acc: 0.8377
Epoch 12/15
20000/20000 - 196s - loss: 0.3798 - acc: 0.8405
Epoch 13/15
20000/20000 - 195s - loss: 0.3708 - acc: 0.8446
Epoch 14/15
20000/20000 - 196s - loss: 0.3626 - acc: 0.8479
Epoch 15/15
20000/20000 - 197s - loss: 0.3551 - acc: 0.8508


In [0]:
# GRU
lo, ac = model3.evaluate(X_train1.astype(np.float32), Y_train)
print("Train Loss: {}".format(lo))
print("Train Accuracy: {}".format(ac)) 

lo, ac = model3.evaluate(X_val1.astype(np.float32), Y_val)
print("Validation Loss: {}".format(lo))
print("Validation Accuracy: {}".format(ac)) 

lo, ac = model3.evaluate(X_test1.astype(np.float32), Y_test)
print("Test Loss: {}".format(lo))
print("Test Accuracy: {}".format(ac)) 

20000/20000 [==============================] - 47s 2ms/sample - loss: 0.3444 - acc: 0.8588
Train Loss: 0.3443882711172104
Train Accuracy: 0.8587999939918518
5000/5000 [==============================] - 12s 2ms/sample - loss: 0.9301 - acc: 0.7506
Validation Loss: 0.9300792888641357
Validation Accuracy: 0.7505999803543091
25000/25000 [==============================] - 59s 2ms/sample - loss: 0.6468 - acc: 0.7612
Test Loss: 0.646802168560028
Test Accuracy: 0.7611600160598755


### (h) Which is the best model among model1 to model3 - RNN, LSTM or GRU? Why? -->

LSTM is the best model as it gives the highest test and validation accuracy. Moreover it has maximum no. of parameters, so in a way has the most complex network which implies more complex learning under the given set of hyperparameters.

### (i) Model 4 -->

In [0]:
#We used LSTM model as it worked the best.

model4 = tf.keras.Sequential([
    Embedding(5000 ,128 , input_length = 200),
    LSTM(200, activation="tanh", return_sequences=True),   #important
    LSTM(200, activation="tanh"),
    Dense(1, activation='sigmoid')
])
print (model4.summary())

Model: "sequential_13"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_16 (Embedding)     (None, 200, 128)          640000    
_________________________________________________________________
lstm_5 (LSTM)                (None, 200, 200)          263200    
_________________________________________________________________
lstm_6 (LSTM)                (None, 200)               320800    
_________________________________________________________________
dense_13 (Dense)             (None, 1)                 201       
Total params: 1,224,201
Trainable params: 1,224,201
Non-trainable params: 0
_________________________________________________________________
None


In [0]:
iter=15
opt = tf.keras.optimizers.Adam(0.0002)
model4.compile(loss='binary_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])

epoc4 = model4.fit(X_train1.astype(np.float32), Y_train, epochs = iter, verbose=2)

Train on 20000 samples
Epoch 1/15
20000/20000 - 447s - loss: 0.4234 - acc: 0.7937
Epoch 2/15
20000/20000 - 446s - loss: 0.2669 - acc: 0.8935
Epoch 3/15
20000/20000 - 439s - loss: 0.2255 - acc: 0.9121
Epoch 4/15
20000/20000 - 447s - loss: 0.1949 - acc: 0.9251
Epoch 5/15
20000/20000 - 449s - loss: 0.1753 - acc: 0.9342
Epoch 6/15
20000/20000 - 448s - loss: 0.1491 - acc: 0.9456
Epoch 7/15
20000/20000 - 447s - loss: 0.1246 - acc: 0.9565
Epoch 8/15
20000/20000 - 453s - loss: 0.1000 - acc: 0.9664
Epoch 9/15
20000/20000 - 452s - loss: 0.0760 - acc: 0.9755
Epoch 10/15
20000/20000 - 444s - loss: 0.0635 - acc: 0.9801
Epoch 11/15
20000/20000 - 444s - loss: 0.0513 - acc: 0.9846
Epoch 12/15
20000/20000 - 448s - loss: 0.0389 - acc: 0.9876
Epoch 13/15
20000/20000 - 455s - loss: 0.0427 - acc: 0.9858
Epoch 14/15
20000/20000 - 441s - loss: 0.0318 - acc: 0.9895
Epoch 15/15
20000/20000 - 436s - loss: 0.0196 - acc: 0.9952


In [1]:
lo, ac = model4.evaluate(X_train1.astype(np.float32), Y_train)
print("Train Loss: {}".format(lo))
print("Train Accuracy: {}".format(ac)) 

lo, ac = model4.evaluate(X_val1.astype(np.float3|2), Y_val)
print("Validation Loss: {}".format(lo))
print("Validation Accuracy: {}".format(ac)) 

lo, ac = model4.evaluate(X_test1.astype(np.float32), Y_test)
print("Test Loss: {}".format(lo))
print("Test Accuracy: {}".format(ac)) 

NameError: ignored

### (j) Model 5 -->

In [0]:
#We used LSTM model as it worked the best.

model5 = tf.keras.Sequential([
    Embedding(5000 ,128 , input_length = 200),
    LSTM(200, activation="tanh", return_sequences=True),   #important
    LSTM(200, activation="tanh", return_sequences=True),
    LSTM(200, activation="tanh"),
    Dense(1, activation='sigmoid')
])
print (model5.summary())

In [0]:
iter=15
opt = tf.keras.optimizers.Adam(0.0002)
model5.compile(loss='binary_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])

epoc5 = model5.fit(X_train1.astype(np.float32), Y_train, epochs = iter, verbose=2)

In [0]:
lo, ac = model5.evaluate(X_train1.astype(np.float32), Y_train)
print("Train Loss: {}".format(lo))
print("Train Accuracy: {}".format(ac)) 

lo, ac = model5.evaluate(X_val1.astype(np.float32), Y_val)
print("Validation Loss: {}".format(lo))
print("Validation Accuracy: {}".format(ac)) 

lo, ac = model5.evaluate(X_test1.astype(np.float32), Y_test)
print("Test Loss: {}".format(lo))
print("Test Accuracy: {}".format(ac)) 

### (k) Model 6 -->

In [0]:
from gensim.models import Word2Vec

In [0]:
# load training data
sentences = X_train
 
# train word2vec model
model6 = Word2Vec(sentences, size=128, window=5, workers=4, min_count=1)
# summarize vocabulary size in model
words = list(model6.wv.vocab)
print('Vocabulary size: %d' % len(words))
 
# save model in ASCII (word2vec) format
filename = 'embedding_word2vec.txt'
model6.wv.save_word2vec_format(filename, binary=False)

### (l) Plot the loss vs iteration and accuracy vs iteration curve for training data for all the models -->

In [0]:
def plotModel(i, ep):
  accuracy_lst = ep.history['acc']
  loss_lst = ep.history['loss']

  print("////////////////////////////////////////// Model {} //////////////////////////////////////////".format(i))

  plt.subplot(121)
  plt.plot(accuracy_lst, c='g')
  plt.xlabel('#iterations')
  plt.ylabel('Accuracy')
  plt.title("Accuracy Plot - Model {} ".format(i))

  plt.subplot(122)    
  plt.plot(loss_lst, c='r')
  plt.xlabel('#iterations')
  plt.ylabel('Loss')
  plt.title("Loss Plot - Model {} ".format(i))

  plt.subplots_adjust(right=1.7)
  plt.show()

  print("//////////////////////////////////////////////////////////////////////////////////////////////")

E = [epoc1,epoc2,epoc3,epoc4,epoc5,epoc6]
for i in range(1,7):
  plotModel(i, E[i-1])

### (m) Tabulate the training, testing and validation accuracy for all the models -->

///////////////|//////////////// ACCURACY /////////////////////////<br>
|- Model ---|--- Training ---|------ Test -----|--- Validation -|<br>
|//////////////|//////////////////////////////////////////////////////////|<br>
|------ 1 ------|--- 100.0 % ---|--- 83.52 % ---|---- 83.20% ----|<br>
|------ 2 ------|--- 99.37 % ---|--- 84.84 % ---|---- 85.96% ----|<br>
|------ 3 ------|--- 85.88 % ---|--- 76.12 % ---|---- 75.06% ----|<br>
|------ 4 ------|--- 99.52 % ---|--- 00.00 % ---|---- 00.00% ----|<br>
|------ 5 ------|--- 00.00 % ---|--- 00.00 % ---|---- 00.00% ----|<br>
|------ 6 ------|--- 00.00 % ---|--- 00.00 % ---|---- 00.00% ----|<br>
////////////////////////////////////////////////////////////////////////////